src/main/java/com/allenai/ml/linalg/DenseVector.java

```diff
@@ -0,0 +1,64 @@
+package com.allenai.ml.linalg;
+
+import lombok.val;
+
+import java.util.Arrays;
+
+public class DenseVector implements Vector {
```

Comment:
There are a bunch and it depends specifically on what you want to do. In general, I trust my numerical intuitions enough to not want to rely on them for trivial things like what’s here (dot products, affine transformations, etc.); if I needed something more complex, like an SVD solver, I’d probably investigate for one. 


The reality is the vast majority of the numerical and ML open-source libs aren’t very good, because very few people have numerical or ML systems in production and so the open-source libraries are primarily for hobbyists or researchers. This is why I probably won’t use any of the existing open-source ML libs. Contrast this with something like Lucene/Netty/etc that is open-source and used heavily in production. 


—
http://aria42.com

On Thu, May 7, 2015 at 8:40 PM, cristipp <notifications@github.com> wrote:

>> @@ -0,0 +1,64 @@
>> +package com.allenai.ml.linalg;
>> +
>> +import lombok.val;
>> +
>> +import java.util.Arrays;
>> +
>> +public class DenseVector implements Vector {
> Fair enough. I'm genuinely interested in a good numeric library for JVM,
> but have little experience in the area. I've used octave and torch, which
> both are Matlab clones (to different degrees). But I don't know of a good
> one for JVM, and the list of options is overwhelming. We're married to the
> JVM, we should have a comprehensive and performant numeric library to use.
> On Thu, May 7, 2015 at 3:04 PM, Aria Haghighi <notifications@github.com>
> wrote:
>> In src/main/java/com/allenai/ml/linalg/DenseVector.java
>> <https://github.com/allenai/ml/pull/2#discussion_r29900562>:
>>
>> > @@ -0,0 +1,64 @@
>> > +package com.allenai.ml.linalg;
>> > +
>> > +import lombok.val;
>> > +
>> > +import java.util.Arrays;
>> > +
>> > +public class DenseVector implements Vector {
>>
>> This PR is primarily about having an interface for vectors I know I need
>> for ML and basic implementations which really wrap around 3rd parties. Even
>> if the implementation that ends up being best is someone else's I'd end up
>> doing an adapter class. The reason for that is that I strongly believe you
>> want your interfaces in terms of library internal classes or JDK class.
>>
>> So I'm sure if performance became an issue with these choices, I'd do an
>> implementation of the interface with those libraries. For the specific case
>> of 1D vectors, I don't think it really matters. The links you have are
>> really for matrix operations.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/allenai/ml/pull/2/files#r29900562>.
>>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/allenai/ml/pull/2/files#r29915306